### ISSUE TITLE
[QST] Extracting Item Embeddings from Trained Model.

### ISSUE THREAD

**Issue by dhollier** [NVIDIA-Merlin/Transformers4Rec#679]
# ❓ Questions & Help

I'm looking to extract the item embeddings for each item after training the model. 

What is the best way to do this?

Digging through past issues I found this code:

`pretrained_embeddings = model.heads[0].body[0].item_embedding_table.weight.cpu().detach().numpy()` ([ #425](https://github.com/NVIDIA-Merlin/Transformers4Rec/issues/425))

However the issue I have with this is that it returns an array of embeddings, but I'm not sure how to match the corresponding Item-Ids.

**Comment by murali-munna** [comment]
I think, the `categories/unique.item_id.parquet` file contains the dictionary that maps the actual item_ids to the item indices used by the model. Using this mapping you can know the item-embeddings correspondence.

**Comment by rnyak** [comment]
@dhollier hello. Yes, as @murali-munna wrote, you can use the unique.item_id.parquet to do the mapping to the original item_ids but couple things to keep in mind:

1. Say you got  [501, 64] from `model.heads[0].body[0].item_embedding_table.weight.cpu().detach().numpy().shape` .  that means, you have 500 unique items but you have +1 outputted and this +1 corresponds to the OOV/nulls. You won’t have OOVs in your train set but you might have in your valid set, and they are mapped to 0 when we apply Categorify() op on validation set.  Note that we also have padding value in TF4Rec model, and it is 0 by default. 

So what you get basically is the embedding weights for encoded item_ids in this order [0, 1, 2, ..., 500]. To convert this to the original item_ids is straightforward. Just use the `categories/unique.item_id.parquet` file and do the mapping.


2. If we use `start_index =1` in the Categorify() op then the item_id encoding starts from 2 not 1. Because we reserve 0 for padding, 1 for OOV +nulls, and then the other item categories start from 2. In that case, you will get 2 + n_unique_items from `model.heads[0].body[0].item_embedding_table.weight.cpu().detach().numpy()` . Meaning,  you will get embedding values for 2 + n_unique_items, where first two are for padding and OOV+nulls.  One caveat here is that our `unique.item_id` parquet file (mapping file) does not reflect the shifted indices when we set `start_index > 0` in the Categorify. You need to shift the index accordingly. You only need to shift the index of the parquet file +1, if you se the start_index =1.

you can check this ticket for further info:

https://github.com/NVIDIA-Merlin/Transformers4Rec/issues/612

**Comment by dhollier** [comment]
Hi @murali-munna & @rnyak,

Thanks for the explanation. This worked!

**Comment by rnyak** [comment]
@dhollier if you dont have further questions, can we close this ticket? thanks.

**Comment by dhollier** [comment]
Yes no further questions this is safe to close 👍
